### PR TITLE
[CALCITE] Disable validation rewrite for merge

### PR DESCRIPTION
**Pull request by ArthurInTheShell**
This pull request is for disabling the validation rewrite for the insert block in a merge call in BigQuery dialect specifically.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java [hunk 1]
```diff
@@ -460,4 +460,28 @@
...
+   * USING t2 AS b
+   * ON a.x = b.x
+   * WHEN MATCHED THEN UPDATE SET
+   *   y = b.y
+   * WHEN NOT MATCHED THEN INSERT (x, y) VALUES (b.x, b.y)
+   * </pre></blockquote>
+   *
+   * <p>{@code VALUES (b.x, b.y)} is the inserted value. The default behavior
+   * is to rewrite it to {@code SELECT b.x, b.y FROM t2 AS b}. Which is not
+   * a valid grammar in BigQuery.
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java:479]: nit: Should we replace "BigQuery" in this comment with "some dialects"?

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 2]
```diff
@@ -1507,7 +1507,7 @@ private void rewriteMerge(SqlMerge call) {
     // note that the values clause has already been converted to a
     // select on the values row constructor; so we need to extract
     // that via the from clause on the select
-    if (insertCall != null) {
+    if (insertCall != null && config.sqlConformance().allowMergeInsertRewrite()) {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:1510]: nit: line wrap

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 3]
```diff
@@ -4901,7 +4901,7 @@ public void validateMerge(SqlMerge call) {
     if (call.getUpdateCall() != null) {
       validateUpdate(call.getUpdateCall());
     }
-    if (call.getInsertCall() != null) {
+    if (call.getInsertCall() != null && config.sqlConformance().allowMergeInsertRewrite()) {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:4904]: nit: line wrap

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java (hunk 1 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java]: Fix in https://github.com/googleinterns/calcite/pull/139/commits/20a224efcc552eb936868ade31f0f9883ea844f6

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 2 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Fix in https://github.com/googleinterns/calcite/pull/139/commits/20a224efcc552eb936868ade31f0f9883ea844f6

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 3 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Fix in https://github.com/googleinterns/calcite/pull/139/commits/20a224efcc552eb936868ade31f0f9883ea844f6

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 4]
```diff
@@ -1507,7 +1507,8 @@ private void rewriteMerge(SqlMerge call) {
     // note that the values clause has already been converted to a
     // select on the values row constructor; so we need to extract
     // that via the from clause on the select
-    if (insertCall != null && config.sqlConformance().allowMergeInsertRewrite()) {
+    if (insertCall != null &&
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:1510]: It looks like the checkstyle test fails unless "&&" is moved to the next line.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 5]
```diff
@@ -4901,7 +4902,8 @@ public void validateMerge(SqlMerge call) {
     if (call.getUpdateCall() != null) {
       validateUpdate(call.getUpdateCall());
     }
-    if (call.getInsertCall() != null && config.sqlConformance().allowMergeInsertRewrite()) {
+    if (call.getInsertCall() != null &&
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:4905]: It looks like the checkstyle test fails unless "&&" is moved to the next line.

Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Fix in https://github.com/googleinterns/calcite/pull/139/commits/eac9ee75ca51ce9dbc1efd141b1bfbe8ecb98e1a

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 4 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Fix in https://github.com/googleinterns/calcite/pull/139/commits/eac9ee75ca51ce9dbc1efd141b1bfbe8ecb98e1a

**Comment by ArthurInTheShell** [comment]
> Can we add 3 unit tests in SqlValidatorTest.java? One which checks that a MERGE statement is no longer rewritten in BIG_QUERY, one which checks that it no longer rewritten in BABEL, and one which checks that it _is_ still rewritten in the default case.
> 
> We should be able to use syntax similar to the following to exercise these different test cases:
> 
> sql(sql).withConformance(SqlConformanceEnum.BABEL).rewritesTo(expected);

Tests are added in https://github.com/googleinterns/calcite/pull/139/commits/02a33237f635fdf08ea7709048dde623b369cee0. These tests are long because we need to use the pre-built table emp and dept for test in this test.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 5 above)
Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Somehow the Github action fails due to the "&&" operator even though changes are made.

Comment (dasch-1) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: It's odd that the GitHub checks are still failing. When I run the autostyle checks locally on your branch, they pass. I've re-triggered the GitHub checks. Hopefully, they'll pass this time around.

Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Re-run all the jobs does not solve the issue. I believe it is some Github action bug. Pushing another commit might help refreshing the cached source code in Github and end up passing the test. I might try that if this run still fails.

Comment (dasch-1) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: It looks like it still fails. Digging into the GitHub test logs, it's definitely still referring to an older version of your code prior to moving the "&&" to a new line.

https://pipelines.actions.githubusercontent.com/xb7m9AKJAufNSYxCgYlBFp6zEZTsNRmGjs0J0ZIVck5oyJLlrb/_apis/pipelines/1/runs/1080/signedlogcontent/6?urlExpires=2020-07-07T16%3A24%3A18.1108720Z&urlSigningMethod=HMACV1&urlSignature=J6TJuli5w8S%2F7Nsv7%2F%2F2GvEQWPJwVMZXZCOTl6TuVTs%3D

Comment (dasch-1) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: The logs are specifically calling that the tests are running on a commit from 7/6 @ 18:05:

2020-07-07T16:13:55.6345088Z [command]/usr/bin/git log -1
2020-07-07T16:13:55.6383774Z commit 83261a192e8ecf5432f95e115b88051bad07d152
2020-07-07T16:13:55.6427770Z Merge: 368ba6f 02a3323
2020-07-07T16:13:55.6427944Z Author: Junhao Chen <32620957+ArthurInTheShell@users.noreply.github.com>
2020-07-07T16:13:55.6428466Z Date:   Mon Jul 6 18:05:21 2020 -0700

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java [hunk 6]
```diff
@@ -460,4 +460,28 @@
...
+   * </pre></blockquote>
+   *
+   * <p>{@code VALUES (b.x, b.y)} is the inserted value. The default behavior
+   * is to rewrite it to {@code SELECT b.x, b.y FROM t2 AS b}. Which is not
+   * a valid grammar in some dialects.
+   *
+   * <p>Among the built-in conformance levels, false in
+   * {@link SqlConformanceEnum#BABEL},
+   * {@link SqlConformanceEnum#BIG_QUERY},
+   * true otherwise.
```
Comment (dstekol) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java:484]: nit: missing \</p> closing block corresponding to \<p> on line 481. Same for \<p> tags on lines 477 and 466.

Comment (ArthurInTheShell) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java]: Good catch! Thanks for pointing it out. Fix in https://github.com/googleinterns/calcite/pull/139/commits/825b67fa1fb053d5c6c0edaffd9d20f3c284b553

**Pull request merged**